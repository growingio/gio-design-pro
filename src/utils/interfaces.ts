import React from 'react';

export interface BaseProps {
  /**
   * 自定义样式名
   */
  className?: string;
  /**
   * 自定义样式属性
   */
  style?: React.CSSProperties;
}

export interface Resource {
  id: string;
  name: string;
  creatorId?: string;
  creator?: string;
  createdAt?: string;
}
