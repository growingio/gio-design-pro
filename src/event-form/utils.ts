/* eslint-disable no-prototype-builtins */
/* eslint-disable consistent-return */

/**
 * 表单数据最大长度
 */
export const MAX_VALUE_LENGTH = 30;
/**
 * 描述字段最大长度
 */
export const MAX_DESC_LENGTH = 150;

export interface Kv {
  key: string;
  value: string;
  [K: string]: any;
}
/**
 * 去除字符串字段中前后的空格
 * @param val
 */
export function trim(val: any): any {
  if (!val) {
    return val;
  }

  if (val.trim) {
    return val.trim.call(val);
  }

  if (Array.isArray(val)) {
    return val.map((v) => trim(v));
  }

  if (typeof val === 'object') {
    const obj = {} as any;
    Object.keys(val).forEach((key) => {
      obj[trim(key)] = trim(val[key]);
    });
    return obj;
  }

  return val;
}

export const isEmptyStr = (val: string) => {
  if (!val) {
    return true;
  }
  return val?.trim().length === 0;
};

export const isObject = (o: any) => Object.prototype.toString.call(o) === '[object Object]';

export const isEmptyValue = (val: any) => val === null || val === undefined || val === '';

export const kvsToQuery = (kvs: Kv[]): any => {
  if (!kvs || kvs.length === 0) return;
  return kvs
    .map((kv) => {
      if (kv?.key) {
        return `${kv?.key}=${kv?.value || ''}`;
      }
      return '';
    })
    .filter((v) => v)
    .join('&');
};

export const objToQuery = (obj: any) =>
  Object.keys(obj || {})
    .map((key) => `${key}=${obj[key]}`)
    .filter((v) => v)
    .join('&');

// 查询参数转化到kv结构数组，过滤掉key不存在的
export const queryToKvs = (query?: string): Kv[] => {
  if (!query) return [];
  // eslint-disable-next-line no-param-reassign
  if (query.startsWith('?')) query = query.slice(1);
  return query
    .split('&')
    .map((kv) => {
      const kva = kv.split('=');
      return {
        key: kva[0],
        value: kva[1],
      };
    })
    .filter((v) => !!v.key);
};
export type Query = { [name: string]: string | any[] };
export const qsParse = (search: string): Query => {
  let _search = search;
  if (search?.startsWith('?')) _search = search.slice(1);
  const query = {} as any;
  _search?.split('&').forEach((item) => {
    const kv = item.split('=');
    if (query.hasOwnProperty(kv[0])) {
      let v = query[kv[0]];
      if (Array.isArray(v)) {
        v.push(kv[1]);
      } else {
        v = [v, kv[1]];
      }
      query[kv[0]] = v?.sort();
    } else {
      !isEmptyStr(kv[0]) && (query[kv[0]] = kv[1] || '');
    }
  });
  return query;
};
/**
 * 判断某个元素是不是一个元素的子元素
 * @param elem 目标元素
 * @param parentElem 父元素
 */
export const isChildOfElement = (elem: Element, parentElem: Element) => {
  let p = elem.parentElement;
  while (p && p !== document.body) {
    if (p === parentElem) {
      return true;
    }
    p = p.parentElement;
  }
  return false;
};
export const matchString = (likePattern: string, value: string, accurate = false): boolean => {
  if (!accurate && (!likePattern || value == null)) {
    return false;
  }
  return !accurate && likePattern.indexOf('*') !== -1
    ? new RegExp(`^${likePattern.replace(/\*+/g, '(.*)')}$`).test(value)
    : likePattern === value;
};
/**
 * 模糊模式下只要page中包含tag中的所有query就行，多了是可以匹配的
 * a=1&b=1可以匹配上a=1
 * 不关注顺序
 *
 * @param patternSearch 匹配模式
 * @param pageSearch 要匹配的query
 * @param accurate 开启精准匹配，参数key一致，值相同，默认false
 */
export const matchQuery = (patternSearch: string, pageSearch: string, accurate = false): boolean => {
  const patternQuery = qsParse(patternSearch);
  const pageQuery = qsParse(pageSearch);

  // 精准匹配下，query个数必须一样
  if (accurate && Object.keys(patternQuery).length !== Object.keys(pageQuery).length) {
    return false;
  }
  // page中的query在多的情况下也可匹配成功
  return Object.keys(patternQuery).every((key) => {
    // eslint-disable-next-line no-prototype-builtins
    if (!pageQuery.hasOwnProperty(key)) {
      return false;
    }
    const pageVal = pageQuery[key] as any;
    const patternVal = patternQuery[key];
    // 匹配规则和页面参数值类型不一样就直接匹配失败,数组和字符串
    if (typeof patternVal !== typeof pageVal) {
      return false;
    }
    if (Array.isArray(patternVal)) {
      // 在精准匹配模式下，数组内容必须是一样的
      if (Array.isArray(pageVal) && (accurate ? pageVal.length === patternVal.length : true)) {
        // 对于数组的参数，比较时不分顺序进行匹配
        return patternVal.every((ptv) => Array.from(pageVal).some((pgv: any) => matchString(ptv, pgv, accurate)));
      }
      return false;
    }
    return matchString(patternVal, pageVal, accurate);
  });
};
