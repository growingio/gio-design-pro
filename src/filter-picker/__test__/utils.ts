/* eslint-disable import/prefer-default-export */
import { act } from 'react-dom/test-utils';

const globalTimeout = global.setTimeout;

export const sleep = async (timeout: number) => {
  await act(async () => {
    await new Promise((resolve) => globalTimeout(resolve, timeout));
  });
};
