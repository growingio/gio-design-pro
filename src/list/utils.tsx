/* eslint-disable import/prefer-default-export */
import React from 'react';
import usePrefixCls from '@gio-design/components/es/utils/hooks/use-prefix-cls';
import Item from './Item';
import ExpandItem from './ExpandItem';
import { ListItemProps } from './interfaces';

const DEFAULT_SHOW_ITEMS_COUNT = 10;

export const rootPrefixCls = () => usePrefixCls('list');

export function renderItem(i: ListItemProps, idx: number) {
  return <Item {...i} key={`item-${i.key}-${idx}`} />;
}

export function renderExpandableItems(expanded: boolean, currentItems: ListItemProps[], onExpand: () => void) {
  if (expanded) {
    return currentItems.map(renderItem);
  }
  if (currentItems.length > DEFAULT_SHOW_ITEMS_COUNT) {
    const showItems = currentItems.slice(0, DEFAULT_SHOW_ITEMS_COUNT);
    return showItems
      .map(renderItem)
      .concat(
        <ExpandItem
          title={`展开全部 (${currentItems.length - showItems.length})`}
          key={`expand-item-${showItems.length + 1}`}
          onClick={onExpand}
        />
      );
  }
  return currentItems.map(renderItem);
}
