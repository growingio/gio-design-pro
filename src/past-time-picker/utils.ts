import has from 'lodash/has';
import get from 'lodash/get';
import { format, startOfToday, sub } from 'date-fns';
import { TimeMode } from './interfaces';
import { DATE_FORMAT } from './constant';

export const parseTimeMode = (timeRange: string | undefined, QUICK_MAPPING: any) => {
  if (!timeRange) {
    return undefined;
  }
  if (has(QUICK_MAPPING, timeRange)) {
    return 'quick';
  }
  const items = timeRange.split(':');
  switch (items[0]) {
    case 'since':
      return TimeMode.Since;
    case 'abs':
      return TimeMode.Absolute;
    case 'day':
      return TimeMode.Relative;
    default:
      return undefined;
  }
};

export const parseStartAndEndDate = (timeRange: string | undefined): [Date | undefined, Date | undefined] => {
  if (!timeRange || timeRange.split(':').length !== 2) {
    return [undefined, undefined];
  }
  const items = timeRange.split(':');
  const times = items[1].split(',').map((str) => parseInt(str, 10));
  const today = startOfToday();
  if (items[0] === 'since') {
    if (times.length === 1) {
      return [new Date(times[0]), today];
    }
    return [new Date(times[0]), sub(today, { days: times[1] })];
  }
  if (items[0] === 'abs') {
    return [new Date(times[0]), new Date(times[1])];
  }
  if (items[0] === 'day') {
    return [sub(today, { days: times[0] }), sub(today, { days: times[1] })];
  }
  return [undefined, undefined];
};

export const parseFixedMode = (timeRange: string | undefined) => {
  if (!timeRange || timeRange.split(':').length !== 2) {
    return false;
  }
  const items = timeRange.split(':');
  const times = items[1].split(',');
  if (items[0] === 'since') {
    if (times.length === 1 || times[1] === '0') {
      return 'today';
    }
    return 'yesterday';
  }
  if (items[0] === 'day' && times[1] === '1') {
    return 'yesterday';
  }
  return false;
};

export const humanizeTimeRange = (
  timeRange: string,
  defaultString: string = '时间范围',
  QUICK_MAPPING: any,
  locale: any
) => {
  const { FromText, toTodayText, toYesterdayText, lastText, dayText, ToText } = locale;
  if (!timeRange || timeRange.split(':').length !== 2) {
    return defaultString;
  }
  if (has(QUICK_MAPPING, timeRange)) {
    return get(QUICK_MAPPING, timeRange);
  }
  const items = timeRange.split(':');
  const times = items[1].split(',').map((str) => parseInt(str, 10));
  if (items[0] === 'since') {
    const start = format(times[0], DATE_FORMAT);
    if (times.length === 1) {
      return `${FromText} ${start} ${toTodayText}`;
    }
    return `${FromText} ${start} ${toYesterdayText}`;
  }
  if (items[0] === 'abs') {
    const start = format(times[0], DATE_FORMAT);
    const end = format(times[1], DATE_FORMAT);
    return `${FromText} ${start} ${ToText} ${end}`;
  }
  if (items[0] === 'day') {
    if (times[1] === 1) {
      return `${lastText} ${times[0] - times[1]} ${dayText}`;
    }
    return `${lastText} ${times[1]}-${times[0]} ${dayText}`;
  }
  return defaultString;
};

export default {
  parseTimeMode,
  parseStartAndEndDate,
  humanizeTimeRange,
};
