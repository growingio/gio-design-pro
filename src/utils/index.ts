import React from 'react';
import { Resource } from './interfaces';

export function resourceToItem(
  resource: Resource,
  mouseEvents?: {
    onClick?: (e: React.MouseEvent<HTMLElement>) => void;
    onMouseEnter?: (e: React.MouseEvent<HTMLElement>) => void;
    onMouseLeave?: (e: React.MouseEvent<HTMLElement>) => void;
  },
  disabled?: boolean
) {
  return {
    key: resource.id,
    children: resource.name,
    ...mouseEvents,
    disabled,
  };
}

export default {
  resourceToItem,
};
