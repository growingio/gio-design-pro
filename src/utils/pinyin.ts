// let allPinyin = [];
// let notone = {};
// let storage = {};

// function init(dict) {
//   allPinyin = Object.keys(dict);
//   notone = parseDict(dict);
//   return match;
// }

// function parseDict(dict) {
//   const parseDict = {};
//   for (const i in dict) {
//     const temp = dict[i];
//     for (let j = 0, len = temp.length; j < len; j++) {
//       if (!parseDict[temp[j]]) {
//         parseDict[temp[j]] = i;
//       } else {
//         parseDict[temp[j]] = `${parseDict[temp[j]]} ${i}`;
//       }
//     }
//   }
//   return parseDict;
// }

// function getPinyin(cn) {
//   const result = [];
//   for (let i = 0, len = cn.length; i < len; i++) {
//     const temp = cn.charAt(i);
//     result.push(notone[temp] || temp);
//   }
//   return result;
// }
// // 对输入拼音进行切分
// function wordBreak(s) {
//   const result = [];
//   const solutions = [];
//   const len = s.length;
//   const possible = [];
//   for (let i = 0; i <= s.length; i++) {
//     possible.push(true);
//   }
//   getAllSolutions(0, s, result, solutions, possible);
//   return solutions;
// }

// function getAllSolutions(start, s, result, solutions, possible) {
//   if (start === s.length) {
//     solutions.push(result.join(' '));
//     return;
//   }
//   for (let i = start; i < s.length; i++) {
//     const piece = s.substring(start, i + 1);
//     let match = false;
//     // 最后一个音特殊处理，不需要全部打完整
//     if (allPinyin.some((i) => i.indexOf(piece) === 0) && !s[i + 1] && possible[i + 1]) {
//       if (piece.length === 1) {
//         result.push(piece);
//       } else {
//         const s = [];
//         allPinyin.forEach((i) => {
//           if (i.indexOf(piece) === 0) {
//             s.push(i);
//           }
//         });
//         result.push(s);
//       }
//       match = true;
//     } else if (allPinyin.indexOf(piece) !== -1 && possible[i + 1]) {
//       result.push(piece);
//       match = true;
//     }
//     if (match) {
//       const beforeChange = solutions.length;
//       getAllSolutions(i + 1, s, result, solutions, possible);
//       if (solutions.length === beforeChange) {
//         possible[i + 1] = false;
//       }
//       result.pop();
//     }
//   }
// }
// // 获取输入拼音的所有组合（切分 + 首字母）
// function getFullKey(key) {
//   const result = [];
//   wordBreak(key).forEach((i) => {
//     const item = i.split(' ');
//     const last = item.length - 1;
//     if (item[last].indexOf(',')) {
//       const keys = item[last].split(',');
//       keys.forEach((j) => {
//         item.splice(last, 1, j);
//         result.push(JSON.parse(JSON.stringify(item)));
//       });
//     } else {
//       result.push(item);
//     }
//   });
//   if (result.length === 0 || result[0].length !== key.length) {
//     result.push(key.split(''));
//   }
//   // 缓存当前结果 避免重复计算
//   storage = { [key]: result };
//   return result;
// }
// function point2point(test, key, last, extend) {
//   if (!test) return false;
//   const a = test.split(' ');
//   a.forEach((i) => {
//     if (i.length > 0 && extend) {
//       a.push(i.charAt(0));
//     }
//   });
//   if (!last) {
//     return a.indexOf(key) !== -1;
//   }
//   return a.some((i) => i.indexOf(key) === 0);
// }

// function match(input, keys) {
//   if (!input || !keys) return false;
//   input = input.toLowerCase();
//   keys = keys.replace(/\s+/g, '').toLowerCase();
//   const indexOf = input.indexOf(keys);
//   if (indexOf !== -1) {
//     return [indexOf, indexOf + keys.length - 1];
//   }
//   // 原文匹配(带空格)
//   const noPyIndex = getIndex(input.split(''), [keys.split('')], keys);
//   if (noPyIndex) return noPyIndex;
//   // pinyin匹配
//   const py = getPinyin(input);
//   const fullString = storage[keys] || getFullKey(keys);
//   return getIndex(py, fullString, keys);
// }
// function getIndex(py, fullString, keys) {
//   for (let p = 0; p < py.length; p++) {
//     for (let k = 0; k < fullString.length; k++) {
//       const key = fullString[k];
//       const keyLength = key.length;
//       const extend = keyLength === keys.length;
//       let isMatch = true;
//       let i = 0;
//       let preSpaceNum = 0;
//       let spaceNum = 0;
//       if (keyLength <= py.length) {
//         for (; i < key.length; i++) {
//           if (i === 0 && py[p + i + preSpaceNum] === ' ') {
//             preSpaceNum += 1;
//             i -= 1;
//           } else if (py[p + i + spaceNum] === ' ') {
//             spaceNum += 1;
//             i -= 1;
//           } else if (!point2point(py[p + i + spaceNum], key[i], !(py[p + i + 1] && key[i + 1]), extend)) {
//             isMatch = false;
//             break;
//           }
//         }
//         if (isMatch) {
//           return [p + preSpaceNum, spaceNum + p + i - 1];
//         }
//       }
//     }
//   }
//   return false;
// }
// export default init;
