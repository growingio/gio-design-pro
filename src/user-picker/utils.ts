import { Option } from '@gio-design/components/es/components/list/interface';
import { Segment } from '../types';
import { OptionGroup, PreparedSegment } from './interfaces';

export function segmentToOption(seg: PreparedSegment | Segment, group?: OptionGroup): Option {
  if (!group) {
    return {
      value: seg.id,
      label: seg.name,
    };
  }
  return {
    value: seg.id,
    label: seg.name,
    groupLabel: group.label,
    groupValue: group.key,
  };
}

export default {
  segmentToOption,
};
