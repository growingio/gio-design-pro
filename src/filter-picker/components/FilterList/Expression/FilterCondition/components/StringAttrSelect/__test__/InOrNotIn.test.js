import React from 'react';
import { mount } from 'enzyme';
import InOrNotIn from '../InOrNotIn.tsx';
import { sleep } from '../../../../../../../__test__/utils.ts';

describe('<InOrNotIn />', () => {
  const originalOffsetHeight = Object.getOwnPropertyDescriptor(HTMLElement.prototype, 'offsetHeight');
  const originalOffsetWidth = Object.getOwnPropertyDescriptor(HTMLElement.prototype, 'offsetWidth');
  beforeEach(() => {
    // For <List /> use <AutoSizer />.
    // <AutoSizer /> uses offsetWidth and offsetHeight.
    // Jest runs in JSDom which doesn't support measurements APIs.
    // @see https://github.com/testing-library/react-testing-library/issues/353#issuecomment-510046921
    Object.defineProperty(HTMLElement.prototype, 'offsetHeight', { configurable: true, value: 320 });
    Object.defineProperty(HTMLElement.prototype, 'offsetWidth', { configurable: true, value: 160 });
  });

  afterEach(() => {
    Object.defineProperty(HTMLElement.prototype, 'offsetHeight', originalOffsetHeight);
    Object.defineProperty(HTMLElement.prototype, 'offsetWidth', originalOffsetWidth);
  });
  it('render', () => {
    const wrapper = mount(<InOrNotIn />);
    expect(wrapper.render()).toMatchSnapshot();
  });

  it('should be can input.', async () => {
    const wrapper = mount(
      <InOrNotIn
        attrChange={() => {}}
        curryDimensionValueRequest={() => new Promise((resolve) => resolve(['www.GrowingIO.com']))}
        valueType="string"
        exprKey="test_key"
        values={[]}
      />
    );
    await sleep(100);
    wrapper.update();
    wrapper.find('.gio-input > input').simulate('change', { target: { value: 't,s' } });
    wrapper.find('.gio-input > input').simulate('change', { target: { value: 't,' } });
    expect(wrapper.render()).toMatchSnapshot();
  });

  it('should be can select', async () => {
    const wrapper = mount(
      <InOrNotIn
        attrChange={() => {}}
        curryDimensionValueRequest={() => new Promise((resolve) => resolve(['www.GrowingIO.com']))}
        valueType="string"
        exprKey="test_key"
        values={[]}
      />
    );
    await sleep(100);
    wrapper.update();
    wrapper.find('.gio-list-wrapper .gio-select-option').simulate('click');
    wrapper.find('.gio-input > input').simulate('change', { target: { value: '' } });
  });

  it('自有输入', async () => {
    const wrapper = mount(
      <InOrNotIn
        attrChange={() => {}}
        curryDimensionValueRequest={() => new Promise((resolve) => resolve(['www.GrowingIO.com']))}
        valueType="string"
        exprKey="test_key"
        values={[]}
      />
    );
    await sleep(100);
    wrapper.update();
    wrapper.find('.gio-input > input').simulate('change', { target: { value: 'test1,' } });
    wrapper.find('.gio-input > input').simulate('change', { target: { value: 'test1,test2' } });
    await sleep(100);
    wrapper.update();
    wrapper.find('.gio-list-wrapper .gio-select-option').at(0).simulate('click');
    wrapper.find('.gio-list-wrapper .gio-select-option').at(0).simulate('click');
  });
});
