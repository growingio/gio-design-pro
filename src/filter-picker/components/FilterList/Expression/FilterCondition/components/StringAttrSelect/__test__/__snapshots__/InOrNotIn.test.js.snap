// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InOrNotIn /> render 1`] = `
<div
  style="height: 330px;"
>
  <div
    class="gio-input gio-input--middle"
  >
    <input
      class="gio-input__content"
      placeholder="请输入…"
      type="text"
      value=""
    />
  </div>
  <div
    style="width: 100%; height: 100%; display: flex; justify-content: center; align-items: center;"
  >
    <div
      class="gio-loading gio-loading-large"
    >
      <span
        class="gio-loading-strip"
      >
        <span
          class="gio-loading-strip-item"
        />
        <span
          class="gio-loading-strip-item"
        />
        <span
          class="gio-loading-strip-item"
        />
        <span
          class="gio-loading-strip-item"
        />
      </span>
      <span
        class="gio-loading-title gio-loading-title-bottom"
      >
        加载中...
      </span>
    </div>
  </div>
</div>
`;

exports[`<InOrNotIn /> should be can input. 1`] = `
<div
  style="height: 330px;"
>
  <div
    class="gio-input gio-input--middle"
  >
    <input
      class="gio-input__content"
      placeholder="请输入…"
      type="text"
      value="t,"
    />
  </div>
  <div
    style="width: 100%; height: 100%; display: flex; justify-content: center; align-items: center;"
  >
    <div
      class="gio-loading gio-loading-large"
    >
      <span
        class="gio-loading-strip"
      >
        <span
          class="gio-loading-strip-item"
        />
        <span
          class="gio-loading-strip-item"
        />
        <span
          class="gio-loading-strip-item"
        />
        <span
          class="gio-loading-strip-item"
        />
      </span>
      <span
        class="gio-loading-title gio-loading-title-bottom"
      >
        加载中...
      </span>
    </div>
  </div>
</div>
`;
